# AWS CloudTrail Security Analytics

This notebook analyzes AWS CloudTrail logs collected from my cloud security capstone environment.

The workflow includes data inspection, cleaning, time-based feature engineering, security risk flags, operational-noise filtering, risk scoring, validation, and export for Power BI.

## 1. Import Libraries

Pandas is used to load, clean, transform, analyze, and export the CloudTrail dataset.

In [1]:
import pandas as pd

## 2. Load the Raw CloudTrail Dataset

The original CSV file is loaded into a Pandas DataFrame named `df`.

In [2]:
# Load the original CloudTrail CSV file
df = pd.read_csv("01_raw_cloudTrail_data.csv")

# Preview the first five rows
df.head()

,User name,AWS access key,Event time,Event source,Event name,AWS region,Source IP address,User agent,Error code,Resources,Request ID,Event ID,Read-only,Event type,Recipient Account Id,Event category
0,root,ASIA3ARNFVNI64IM3R47,2026-06-19T13:26:02Z,iam.amazonaws.com,AttachUserPolicy,us-east-1,99.237.32.148,Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...,NaN,"[{""resourceType"":""AWS::IAM::User"",""resourceNam...",3b3d2686-8988-4860-8d76-822595bb3460,3111876f-b85e-4673-8397-4a49c3a9aeb3,False,AwsApiCall,757082729297,Management
1,root,ASIA3ARNFVNI64IM3R47,2026-06-19T13:24:26Z,iam.amazonaws.com,RemoveUserFromGroup,us-east-1,99.237.32.148,Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...,NaN,"[{""resourceType"":""AWS::IAM::User"",""resourceNam...",d2f492a0-9dba-4b4c-9b7f-e6120e1f73eb,66281920-7f73-4486-8cac-411e448896a0,False,AwsApiCall,757082729297,Management
2,root,ASIA3ARNFVNI6YYHPDA6,2026-06-19T13:22:52Z,iam.amazonaws.com,EnableMFADevice,us-east-1,99.237.32.148,Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...,NaN,"[{""resourceType"":""AWS::IAM::User"",""resourceNam...",f091add1-8d63-4262-9204-552661998ef2,0bf9d0c8-3059-480c-ad48-799fd7be7c7a,False,AwsApiCall,757082729297,Management
3,root,ASIA3ARNFVNI6YYHPDA6,2026-06-19T13:20:06Z,iam.amazonaws.com,CreateVirtualMFADevice,us-east-1,99.237.32.148,Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...,NaN,"[{""resourceType"":""AWS::IAM::MfaDevice"",""resour...",36f4ae00-eec5-4548-b289-13cfd03a8f9b,3df01fc9-1875-4d6f-b954-0e0cc6c773e0,False,AwsApiCall,757082729297,Management
4,root,NaN,2026-06-19T13:06:52Z,signin.amazonaws.com,ConsoleLogin,us-east-1,99.237.32.148,Mozilla/5.0 (Windows NT 10.0; Win64; x64) Appl...,NaN,[],NaN,db233914-b832-41c2-a4fa-a3ebe62f12e5,False,AwsConsoleSignIn,757082729297,Management


## 3. Inspect the Dataset

Before cleaning the data, I review its size, column names, data types, and missing values.

In [3]:
print("Rows and columns:", df.shape)
print("\nColumn names:")
print(df.columns.tolist())

Rows and columns: (2814, 16)

Column names:
['User name', 'AWS access key', 'Event time', 'Event source', 'Event name', 'AWS region', 'Source IP address', 'User agent', 'Error code', 'Resources', 'Request ID', 'Event ID', 'Read-only', 'Event type', 'Recipient Account Id', 'Event category']


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2814 entries, 0 to 2813
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   User name             2811 non-null   object
 1   AWS access key        2750 non-null   object
 2   Event time            2814 non-null   object
 3   Event source          2814 non-null   object
 4   Event name            2814 non-null   object
 5   AWS region            2814 non-null   object
 6   Source IP address     2814 non-null   object
 7   User agent            2814 non-null   object
 8   Error code            21 non-null     object
 9   Resources             2814 non-null   object
 10  Request ID            2751 non-null   object
 11  Event ID              2814 non-null   object
 12  Read-only             2814 non-null   bool  
 13  Event type            2814 non-null   object
 14  Recipient Account Id  2814 non-null   int64 
 15  Event category        2814 non-null   

### Observation

The dataset contains 2,814 CloudTrail events across 16 columns. Most fields are complete, while some optional fields (such as Error code) contain missing values, which is expected because errors occur only for certain AWS events.

## 4. Create a Working Copy and Clean Column Names

The original DataFrame is preserved. All cleaning and feature engineering are performed on `df_clean`.

In [5]:
df_clean = df.copy()

df_clean.columns = df_clean.columns.str.strip()

print(df_clean.columns.tolist())

['User name', 'AWS access key', 'Event time', 'Event source', 'Event name', 'AWS region', 'Source IP address', 'User agent', 'Error code', 'Resources', 'Request ID', 'Event ID', 'Read-only', 'Event type', 'Recipient Account Id', 'Event category']


## 5. Convert Event Time and Create Time Features

The event timestamp is converted from text into a timezone-aware datetime value. New time fields are then created for dashboard analysis.

In [7]:
df_clean["Event time"] = pd.to_datetime(
    df_clean["Event time"],
    errors="coerce",
    utc=True
)

df_clean["Hour_of_Day"] = df_clean["Event time"].dt.hour
df_clean["Day_Name"] = df_clean["Event time"].dt.day_name()
df_clean["Day_of_Week"] = df_clean["Event time"].dt.dayofweek
df_clean["Is_Weekend"] = df_clean["Day_of_Week"].isin([5, 6])

In [8]:
def get_time_bucket(hour):
    """Assign an hour to a time-of-day category."""
    if 6 <= hour < 12:
        return "Morning"
    elif 12 <= hour < 17:
        return "Afternoon"
    elif 17 <= hour < 21:
        return "Night"
    else:
        return "Outside-Hours"


df_clean["Time_of_Day_Bucket"] = (
    df_clean["Hour_of_Day"].apply(get_time_bucket)
)

In [9]:
df_clean[[
    "Event time",
    "Hour_of_Day",
    "Day_Name",
    "Is_Weekend",
    "Time_of_Day_Bucket"
]].head(10)

,Event time,Hour_of_Day,Day_Name,Is_Weekend,Time_of_Day_Bucket
0,2026-06-19 13:26:02+00:00,13,Friday,False,Afternoon
1,2026-06-19 13:24:26+00:00,13,Friday,False,Afternoon
2,2026-06-19 13:22:52+00:00,13,Friday,False,Afternoon
3,2026-06-19 13:20:06+00:00,13,Friday,False,Afternoon
4,2026-06-19 13:06:52+00:00,13,Friday,False,Afternoon
5,2026-06-15 06:11:10+00:00,6,Monday,False,Morning
6,2026-06-06 04:40:43+00:00,4,Saturday,True,Outside-Hours
7,2026-06-06 02:34:56+00:00,2,Saturday,True,Outside-Hours
8,2026-06-06 01:49:55+00:00,1,Saturday,True,Outside-Hours
9,2026-06-06 01:47:25+00:00,1,Saturday,True,Outside-Hours


## 6. Create Security Flags

Boolean security features are created to identify root-account usage, high-risk AWS actions, console logins, and denied or unauthorized activity.

In [10]:
df_clean["Is_Root_Account"] = (
    df_clean["User name"]
    .fillna("")
    .str.strip()
    .str.lower()
    .eq("root")
)

In [11]:
high_risk_actions = [
    "StopLogging",
    "DeleteTrail",
    "StartInstances",
    "StopInstances",
    "TerminateInstances",
    "AuthorizeSecurityGroupIngress",
    "CreateAccessKey",
    "AttachUserPolicy",
    "PutBucketPolicy",
    "DeleteBucket",
    "CreateVirtualMFADevice"
]

df_clean["Is_High_Risk_Action"] = (
    df_clean["Event name"].isin(high_risk_actions)
)

In [12]:
df_clean["Is_Console_Login"] = (
    df_clean["Event name"].eq("ConsoleLogin")
)

df_clean["Is_Security_Error"] = (
    df_clean["Error code"]
    .fillna("")
    .str.contains(
        r"Denied|Unauthorized|AccessDenied",
        case=False,
        regex=True
    )
)

In [13]:
print("Root events:", df_clean["Is_Root_Account"].sum())
print("High-risk actions:", df_clean["Is_High_Risk_Action"].sum())
print("Console logins:", df_clean["Is_Console_Login"].sum())
print("Security errors:", df_clean["Is_Security_Error"].sum())

Root events: 291
High-risk actions: 83
Console logins: 60
Security errors: 4


## 7. Create the Risk Score

Each event starts with a score of zero. Points are added for root-account activity, high-risk actions, security errors, and activity outside standard hours.

In [14]:
df_clean["Risk_Score"] = 0

df_clean.loc[
    df_clean["Is_Root_Account"],
    "Risk_Score"
] += 5

df_clean.loc[
    df_clean["Is_High_Risk_Action"],
    "Risk_Score"
] += 7

df_clean.loc[
    df_clean["Is_Security_Error"],
    "Risk_Score"
] += 4

df_clean.loc[
    df_clean["Time_of_Day_Bucket"].eq("Outside-Hours"),
    "Risk_Score"
] += 2

### Risk-scoring rules

- Root-account event: +5  
- High-risk AWS action: +7  
- Security error: +4  
- Outside-hours activity: +2  

## 8. Assign Refined Risk Levels

The initial risk distribution was reviewed and the thresholds were refined. Critical is reserved for the strongest combined-risk conditions.

In [15]:
def get_risk_level(score):
    """Convert a numeric risk score into a severity category."""
    if score >= 14:
        return "Critical"
    elif score >= 10:
        return "High"
    elif score >= 5:
        return "Medium"
    else:
        return "Low"


df_clean["Risk_Level"] = (
    df_clean["Risk_Score"].apply(get_risk_level)
)

In [16]:
df_clean["Risk_Level"].value_counts()

,count
Risk_Level,
Low,2515
Medium,221
Critical,48
High,30


## 9. Identify and Filter Operational Noise

Before filtering, the frequency of AWS event names is examined to identify repetitive operational activity that is not relevant to security analysis.

In [17]:
df_clean["Event name"].value_counts().head(10)

,count
Event name,
UpdateInstanceInformation,2461
PutMetricAlarm,71
ConsoleLogin,60
StartInstances,27
StopInstances,26
CreateLogStream,25
RevokeSecurityGroupIngress,18
DetachUserPolicy,18
AuthorizeSecurityGroupIngress,14


### Observation

Exploratory analysis showed that **UpdateInstanceInformation** dominated the dataset. These repetitive AWS Systems Manager heartbeat events represent routine operational activity rather than security incidents.

To preserve the original dataset, these events were flagged as operational noise instead of being permanently removed. The original data remains in **df_clean**, while **df_security** contains the security-focused events used for Power BI analysis.

In [18]:
noise_events = [
    "UpdateInstanceInformation"
]

df_clean["Is_Operational_Noise"] = (
    df_clean["Event name"].isin(noise_events)
)

df_security = df_clean[
    ~df_clean["Is_Operational_Noise"]
].copy()

In [19]:
print("Original events:", len(df_clean))
print(
    "Operational noise events:",
    df_clean["Is_Operational_Noise"].sum()
)
print("Security-focused events:", len(df_security))

Original events: 2814
Operational noise events: 2461
Security-focused events: 353


## 10. Validate the Security-Focused Dataset

The score ranges and event counts are checked to confirm that the risk categories do not overlap and that the final model behaves as intended.

In [20]:
risk_validation = (
    df_security.groupby("Risk_Level")["Risk_Score"]
    .agg(
        Minimum="min",
        Maximum="max",
        Count="count"
    )
)

risk_validation

,Minimum,Maximum,Count
Risk_Level,,,
Critical,14,14,48
High,11,13,30
Low,0,4,54
Medium,5,9,221


In [21]:
df_security["Risk_Score"].value_counts().sort_index()

,count
Risk_Score,
0,13
2,40
4,1
5,75
6,1
7,140
9,5
11,1
12,28


In [ ]:
df_security["Risk_Level"].value_counts()

,count
Risk_Level,
Medium,221
Low,54
Critical,48
High,30


### Validation Result

The refined scoring model produces four non-overlapping severity levels. Critical events require the highest combined risk score, while High, Medium, and Low represent progressively lower security risk.

## 11. Add Power BI Risk-Level Sort Order

A numeric sort column ensures that Power BI displays severity in the correct order: Critical, High, Medium, Low.

In [22]:
risk_order = {
    "Critical": 1,
    "High": 2,
    "Medium": 3,
    "Low": 4
}

df_security["Risk_Sort"] = (
    df_security["Risk_Level"].map(risk_order)
)

In [ ]:
df_security[[
    "Risk_Level",
    "Risk_Sort"
]].drop_duplicates().sort_values("Risk_Sort")

,Risk_Level,Risk_Sort
6,Critical,1
0,High,2
1,Medium,3
343,Low,4


## 12. Preview and Export the Final Dataset

The final security-focused dataset is reviewed and exported for Power BI.

In [23]:
final_columns = [
    "Event time",
    "Time_of_Day_Bucket",
    "User name",
    "Event name",
    "Source IP address",
    "Is_Root_Account",
    "Is_High_Risk_Action",
    "Is_Console_Login",
    "Is_Security_Error",
    "Risk_Score",
    "Risk_Level",
    "Risk_Sort"
]

df_security[final_columns].head(20)

,Event time,Time_of_Day_Bucket,User name,Event name,Source IP address,Is_Root_Account,Is_High_Risk_Action,Is_Console_Login,Is_Security_Error,Risk_Score,Risk_Level,Risk_Sort
0,2026-06-19 13:26:02+00:00,Afternoon,root,AttachUserPolicy,99.237.32.148,True,True,False,False,12,High,2
1,2026-06-19 13:24:26+00:00,Afternoon,root,RemoveUserFromGroup,99.237.32.148,True,False,False,False,5,Medium,3
2,2026-06-19 13:22:52+00:00,Afternoon,root,EnableMFADevice,99.237.32.148,True,False,False,False,5,Medium,3
3,2026-06-19 13:20:06+00:00,Afternoon,root,CreateVirtualMFADevice,99.237.32.148,True,True,False,False,12,High,2
4,2026-06-19 13:06:52+00:00,Afternoon,root,ConsoleLogin,99.237.32.148,True,False,True,False,5,Medium,3
5,2026-06-15 06:11:10+00:00,Morning,root,ConsoleLogin,99.237.32.148,True,False,True,False,5,Medium,3
6,2026-06-06 04:40:43+00:00,Outside-Hours,root,StopInstances,99.237.32.148,True,True,False,False,14,Critical,1
7,2026-06-06 02:34:56+00:00,Outside-Hours,root,StartInstances,99.237.32.148,True,True,False,False,14,Critical,1
8,2026-06-06 01:49:55+00:00,Outside-Hours,root,CreateAccessKey,99.237.32.148,True,True,False,False,14,Critical,1
9,2026-06-06 01:47:25+00:00,Outside-Hours,root,AttachUserPolicy,99.237.32.148,True,True,False,False,14,Critical,1


In [24]:
df_security.to_csv(
    "cloudtrail_security_events.csv",
    index=False
)

print(
    "Export complete:",
    "cloudtrail_security_events.csv"
)

Export complete: cloudtrail_security_events.csv


This dataset will be imported into Power BI to build the Executive Overview and Security Investigation dashboards.

In [25]:
print("Original CloudTrail Events:", len(df_clean))
print("Operational Noise Events:", df_clean["Is_Operational_Noise"].sum())
print("Security Events:", len(df_security))
print("Root Account Events:", df_security["Is_Root_Account"].sum())
print("High-Risk Actions:", df_security["Is_High_Risk_Action"].sum())
print("Console Logins:", df_security["Is_Console_Login"].sum())
print("Security Errors:", df_security["Is_Security_Error"].sum())

Original CloudTrail Events: 2814
Operational Noise Events: 2461
Security Events: 353
Root Account Events: 291
High-Risk Actions: 83
Console Logins: 60
Security Errors: 4


## Final Results

The original CloudTrail dataset contained **2,814** AWS events collected from a cloud security capstone environment.

After identifying and removing **2,461** operational-noise events, the final dataset contained **353** security-focused events for analysis.

### Key Findings

- Original CloudTrail events: **2,814**
- Operational-noise events removed: **2,461**
- Security-focused events: **353**
- Root-account events: **291**
- High-risk AWS actions: **83**
- Console logins: **60**
- Security errors: **4**

### Project Deliverables

- Cleaned CloudTrail security dataset
- Security risk scoring using Python (Pandas)
- Operational-noise filtering
- Executive Overview dashboard (Power BI)
- Security Investigation dashboard (Power BI)

This dataset demonstrates a complete workflow from raw CloudTrail logs to security-focused analytics and interactive dashboard reporting.